app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.java

```diff
@@ -0,0 +1,120 @@
+package com.quran.labs.androidquran.widgets;
+
+import android.appwidget.AppWidgetManager;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.widget.RemoteViews;
+import android.widget.RemoteViewsService;
+
+import com.quran.labs.androidquran.R;
+import com.quran.labs.androidquran.dao.Bookmark;
+import com.quran.labs.androidquran.data.QuranInfo;
+import com.quran.labs.androidquran.database.BookmarksDBAdapter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WidgetListProvider implements RemoteViewsService.RemoteViewsFactory {
+  private ArrayList<ItemList> listItemList = new ArrayList();
+  private Context context = null;
+  private int appWidgetId;
+
+  public WidgetListProvider(Context context, Intent intent) {
+    this.context = context;
+    appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID,
+        AppWidgetManager.INVALID_APPWIDGET_ID);
+    populateListItem();
+  }
+
+  public void populateListItem() {
+    Context appContext = context.getApplicationContext();
+    BookmarksDBAdapter a = new BookmarksDBAdapter(appContext);
+    List<Bookmark> bookmarksList = a.getBookmarks(0);// 0 = date added
+    listItemList = new ArrayList();
+    for (int i = 0; i < bookmarksList.size(); i++) {
+      Bookmark bm = bookmarksList.get(i);
+      ItemList item = new ItemList();
+      item.sura = QuranInfo.getSuraNameFromPage(appContext, bm.page);
+      item.page = bm.page;
+      listItemList.add(item);
+    }
+
+
+  }
+
+  @Override
+  public void onCreate() {
+  }
+
+  @Override
+  public void onDataSetChanged() {
+    populateListItem();
+
+  }
+
+  @Override
+  public void onDestroy() {
+  }
+
+  @Override
+  public int getCount() {
+    return listItemList.size();
+  }
+
+  @Override
+  public long getItemId(int position) {
+    return position;
+  }
+
+  @Override
+  public boolean hasStableIds() {
+    return false;
+  }
+
+  /*
```

Comment:
fix the comment - you can use intellij's code format feature to format it properly